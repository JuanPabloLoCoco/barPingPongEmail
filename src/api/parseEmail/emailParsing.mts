import { load, CheerioAPI } from "cheerio";
import { parsingDate } from "./parseDateTime.mjs";

interface BaseReservation {
  name: string;
  venue: string;
  startDate: Date;
}

export enum ReservationType {
  NEW = "NEW",
  CANCEL = "CANCEL",
}

export interface EmailNewReservation extends BaseReservation {
  type: ReservationType.NEW;
  phone: string;
  endDate: Date;
}

export interface CancelReservation extends BaseReservation {
  type: ReservationType.CANCEL;
}

export type EmailOperation = EmailNewReservation | CancelReservation;

export function emailParsing(
  html: string,
  referenceDate: Date = new Date()
): EmailNewReservation | CancelReservation {
  // Parse the email
  const cheerioAPI: CheerioAPI = load(html);

  // Extract the information
  const venue = cheerioAPI('td:contains("Cancha")').next().text().trim();

  const isNewReservation =
    cheerioAPI('h1:contains("¡Nueva reserva online!")').length > 0;

  if (isNewReservation) {
    return parseNewReservation(cheerioAPI, venue, referenceDate);
  }

  const isCancelReservation = cheerioAPI("body")
    .text()
    .includes("Reserva online cancelada");

  if (!isCancelReservation) {
    throw new Error("Invalid email");
  }

  const name = cheerioAPI('td:contains("Jugador")')
    .next() // Selecciona el siguiente 'td' que contiene el nombre
    .text() // Obtiene el texto del elemento
    .trim(); // Elimina espacios en blanco al principio y al final

  const startDateStr = cheerioAPI('td:contains("Dia y hora")')
    .next()
    .text()
    .trim();

  let startDate: Date | null = null;
  try {
    startDate = parsingDate(startDateStr, referenceDate);
  } catch (err) {
    console.log("Error while parsing date");
  }

  return {
    type: ReservationType.CANCEL,
    name,
    venue,
    startDate: startDate || new Date(),
  };
}

function parseNewReservation(
  $: CheerioAPI,
  venue: string,
  referenceDate: Date
): EmailNewReservation {
  const name = $('td:contains("Nombre")').next().text().trim();
  const phone = $('td:contains("Teléfono")').next().text().trim();
  const startDateStr = $('td:contains("Dia y hora")').next().text().trim();

  // Convertir la fecha a un objeto Date
  let startDate: Date | null = null;
  try {
    startDate = parsingDate(startDateStr, referenceDate);
  } catch (err) {
    console.log("Error while parsing date");
  }

  return {
    type: ReservationType.NEW,
    name,
    venue,
    startDate: startDate || new Date(),
    endDate: new Date(),
    phone,
  };
}
