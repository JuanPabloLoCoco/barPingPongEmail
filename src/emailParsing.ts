import { load } from "cheerio";
import { parsingDate, separarFechaHora } from "./utils/parseDateTime";

interface BaseReservation {
  name: string;
  venue: string;
  startDate: Date;
}

export enum ReservationType {
  NEW = "NEW",
  CANCEL = "CANCEL",
}

export interface NewReservation extends BaseReservation {
  type: ReservationType.NEW;
  phone: string;
  endDate: Date;
}

export interface CancelReservation extends BaseReservation {
  type: ReservationType.CANCEL;
}

export type EventType = NewReservation | CancelReservation;

export function emailParsing(
  html: string,
  referenceDate: Date = new Date()
): NewReservation | CancelReservation {
  // Parse the email
  const $ = load(html);

  // Extract the information
  const venue = $('td:contains("Cancha")').next().text().trim();

  const isNewReservation =
    $('h1:contains("¡Nueva reserva online!")').length > 0;

  if (isNewReservation) {
    return parseNewReservation($, venue, referenceDate);
  }

  const isCancelReservation = $("body")
    .text()
    .includes("Reserva online cancelada");

  if (!isCancelReservation) {
    throw new Error("Invalid email");
  }

  const name = $('td:contains("Jugador")')
    .next() // Selecciona el siguiente 'td' que contiene el nombre
    .text() // Obtiene el texto del elemento
    .trim(); // Elimina espacios en blanco al principio y al final

  const startDateStr = $('td:contains("Dia y hora")').next().text().trim();

  let startDate: Date | null = null;
  try {
    startDate = parsingDate(startDateStr, referenceDate);
  } catch (err) {
    console.log("Error while parsing date");
  }

  return {
    type: ReservationType.CANCEL,
    name,
    venue,
    startDate: startDate || new Date(),
  };
}

function parseNewReservation(
  $: cheerio.Root,
  venue: string,
  referenceDate: Date
): NewReservation {
  const name = $('td:contains("Nombre")').next().text().trim();
  const phone = $('td:contains("Teléfono")').next().text().trim();
  const startDateStr = $('td:contains("Dia y hora")').next().text().trim();

  // Convertir la fecha a un objeto Date
  let startDate: Date | null = null;
  try {
    startDate = parsingDate(startDateStr, referenceDate);
  } catch (err) {
    console.log("Error while parsing date");
  }

  return {
    type: ReservationType.NEW,
    name,
    venue,
    startDate: startDate || new Date(),
    endDate: new Date(),
    phone,
  };
}
