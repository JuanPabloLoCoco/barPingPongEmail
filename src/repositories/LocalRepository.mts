import { NewReservation } from "../emailParsing.mjs";
import {
  CancelledReservation,
  CreatedReservation,
  NotifiedReservation,
  ReservationRepository,
  ReservationState,
  ReservationType,
  ReservationWithToken,
} from "./Reservation.mjs";

function buildId(date: Date, venue: string, name: string): string {
  return `${date.getTime()}_${venue}_${name}`;
}
export class LocalReservationRepository implements ReservationRepository {
  idMap: Map<string, ReservationType> = new Map();

  create(reservation: NewReservation): Promise<CreatedReservation> {
    const createdReservation: CreatedReservation = {
      id: buildId(reservation.startDate, reservation.venue, reservation.name),
      state: ReservationState.CREATED,
      reservation,
      createdAt: new Date(),
      updatedAt: null,
    };
    this.idMap.set(createdReservation.id, createdReservation);
    return Promise.resolve(createdReservation);
  }
  asignToken(
    reservation: CreatedReservation,
    token: string,
    password_id: string
  ): Promise<ReservationWithToken> {
    const reservationWithToken: ReservationWithToken = {
      ...reservation,
      state: ReservationState.WITH_TOKEN,
      token,
      password_id,
    };
    this.idMap.set(reservationWithToken.id, reservationWithToken);
    return Promise.resolve(reservationWithToken);
  }
  get(id: string): Promise<ReservationType | null> {
    const reservationFound = this.idMap.get(id);
    if (!reservationFound) {
      throw new Error("Reservation not found");
    }
    return Promise.resolve(reservationFound);
  }
  findReservation(reservation: {
    date: Date;
    venue: string;
    name: string;
  }): Promise<ReservationType | null> {
    const id = buildId(reservation.date, reservation.venue, reservation.name);
    const reservationFound = this.idMap.get(id);
    return Promise.resolve(reservationFound ?? null);
  }
  cancel(id: string): Promise<CancelledReservation> {
    const reservationFound = this.idMap.get(id);
    if (
      !reservationFound ||
      reservationFound.state === ReservationState.CANCELLED
    ) {
      throw new Error("Reservation not found");
    }
    const cancelledReservation: CancelledReservation = {
      id: reservationFound.id,
      state: ReservationState.CANCELLED,
      reservation: reservationFound.reservation,
      createdAt: reservationFound.createdAt,
      updatedAt: new Date(),
    };
    this.idMap.set(id, cancelledReservation);
    return Promise.resolve(cancelledReservation);
  }
  getFutureReservations(): Promise<NotifiedReservation[]> {
    const today = new Date().getTime();
    const reservationArray = Array.from(this.idMap.values());

    return Promise.resolve(
      reservationArray.filter(
        (x) =>
          x.state === ReservationState.NOTIFIED &&
          x.reservation.startDate.getTime() >= today
      ) as NotifiedReservation[]
    );
  }
  notify(reservation: ReservationWithToken): Promise<NotifiedReservation> {
    const notifiedReservation: NotifiedReservation = {
      ...reservation,
      state: ReservationState.NOTIFIED,
    };
    this.idMap.set(notifiedReservation.id, notifiedReservation);
    return Promise.resolve(notifiedReservation);
  }
}
