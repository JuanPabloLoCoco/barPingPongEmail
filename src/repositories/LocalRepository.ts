import { NewReservation } from "../emailParsing";
import {
  CancelledReservation,
  CreatedReservation,
  NotifiedReservation,
  ReservationRepository,
  ReservationState,
  ReservationType,
  ReservationWithToken,
} from "./Reservation";

export class LocalReservationRepository implements ReservationRepository {
  private idMap: Map<string, ReservationType> = new Map();

  constructor() {
    this.idMap = new Map();
  }

  create(reservation: NewReservation): Promise<CreatedReservation> {
    const createdReservation: CreatedReservation = {
      id: `${reservation.startDate.getTime()}_${reservation.venue}_${
        reservation.name
      }`,
      state: ReservationState.CREATED,
      reservation,
      createdAt: new Date(),
      updatedAt: null,
    };
    this.idMap.set(createdReservation.id, createdReservation);
    return Promise.resolve(createdReservation);
  }
  asignToken(
    reservation: CreatedReservation,
    newToken: string
  ): Promise<ReservationWithToken> {
    const reservationWithToken: ReservationWithToken = {
      ...reservation,
      state: ReservationState.WITH_TOKEN,
      token: newToken,
    };
    this.idMap.set(reservationWithToken.id, reservationWithToken);
    return Promise.resolve(reservationWithToken);
  }
  get(id: string): Promise<ReservationType | null> {
    const reservationFound = this.idMap.get(id);
    if (!reservationFound) {
      throw new Error("Reservation not found");
    }
    return Promise.resolve(reservationFound);
  }
  cancel(id: string): Promise<CancelledReservation> {
    const reservationFound = this.idMap.get(id);
    if (
      !reservationFound ||
      reservationFound.state === ReservationState.CANCELLED
    ) {
      throw new Error("Reservation not found");
    }
    const cancelledReservation: CancelledReservation = {
      id: reservationFound.id,
      state: ReservationState.CANCELLED,
      reservation: reservationFound.reservation,
      createdAt: reservationFound.createdAt,
      updatedAt: new Date(),
    };
    this.idMap.set(id, cancelledReservation);
    return Promise.resolve(cancelledReservation);
  }
  getFutureReservations(): Promise<NotifiedReservation[]> {
    const today = new Date().getTime();
    const reservationArray = Array.from(this.idMap.values());

    return Promise.resolve(
      reservationArray.filter(
        (x) =>
          x.state === ReservationState.NOTIFIED &&
          x.reservation.startDate.getTime() >= today
      ) as NotifiedReservation[]
    );
  }
  notify(reservation: ReservationWithToken): Promise<NotifiedReservation> {
    const notifiedReservation: NotifiedReservation = {
      ...reservation,
      state: ReservationState.NOTIFIED,
    };
    this.idMap.set(notifiedReservation.id, notifiedReservation);
    return Promise.resolve(notifiedReservation);
  }
}
