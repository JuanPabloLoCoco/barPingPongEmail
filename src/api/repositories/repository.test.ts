import { describe, test, expect } from "@jest/globals";
import { LocalReservationRepository } from "./LocalRepository";
import { NewReservation, ReservationType } from "../parseEmail/emailParsing";

describe("LocalRepository test", () => {
  test("Happy case", async () => {
    const repository = new LocalReservationRepository();
    const reservation: NewReservation = {
      startDate: new Date(),
      venue: "venue",
      name: "name",
      type: ReservationType.NEW,
      endDate: new Date(),
      phone: "+5491140405050",
    };

    const createdReservation = await repository.create(reservation);

    expect(createdReservation).toMatchSnapshot({
      id: expect.any(String),
      state: "CREATED",
      reservation,
      createdAt: expect.any(Date),
      updatedAt: null,
    });

    // Get reservation
    const reservationFound = await repository.get(createdReservation.id);
    expect(reservationFound).toEqual(createdReservation);

    // Add token to reservation
    const token = "1234567";

    const reservationWithToken = await repository.asignToken(
      createdReservation,
      token,
      token
    );
    expect(reservationWithToken).toMatchSnapshot({
      ...reservationWithToken,
      id: expect.any(String),
      createdAt: expect.any(Date),
      updatedAt: expect.any(Date),
    });

    const notifyReservation = await repository.notify(reservationWithToken);
    expect(notifyReservation).toMatchSnapshot({
      ...notifyReservation,
      id: expect.any(String),
      createdAt: expect.any(Date),
      updatedAt: expect.any(Date),
    });

    const cancelReservation = await repository.cancel(notifyReservation.id);
    expect(cancelReservation).toMatchSnapshot({
      ...cancelReservation,
      state: "CANCELLED",
      id: expect.any(String),
      createdAt: expect.any(Date),
      updatedAt: expect.any(Date),
    });
  });
});
