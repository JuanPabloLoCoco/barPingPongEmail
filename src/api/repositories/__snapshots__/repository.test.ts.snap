// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LocalRepository test Happy case 1`] = `
{
  "createdAt": Any<Date>,
  "id": Any<String>,
  "reservation": {
    "endDate": {},
    "name": "name",
    "phone": "+5491140405050",
    "startDate": {},
    "type": "NEW",
    "venue": "venue",
  },
  "state": "CREATED",
  "updatedAt": null,
}
`;

exports[`LocalRepository test Happy case 2`] = `
{
  "createdAt": Any<Date>,
  "id": Any<String>,
  "password_id": "1234567",
  "reservation": {
    "endDate": {},
    "name": "name",
    "phone": "+5491140405050",
    "startDate": {},
    "type": "NEW",
    "venue": "venue",
  },
  "state": "WITH_TOKEN",
  "token": "1234567",
  "updatedAt": Any<Date>,
}
`;

exports[`LocalRepository test Happy case 3`] = `
{
  "createdAt": Any<Date>,
  "id": Any<String>,
  "password_id": "1234567",
  "reservation": {
    "endDate": {},
    "name": "name",
    "phone": "+5491140405050",
    "startDate": {},
    "type": "NEW",
    "venue": "venue",
  },
  "state": "NOTIFIED",
  "token": "1234567",
  "updatedAt": Any<Date>,
}
`;

exports[`LocalRepository test Happy case 4`] = `
{
  "createdAt": Any<Date>,
  "id": Any<String>,
  "reservation": {
    "endDate": {},
    "name": "name",
    "phone": "+5491140405050",
    "startDate": {},
    "type": "NEW",
    "venue": "venue",
  },
  "state": "CANCELLED",
  "updatedAt": Any<Date>,
}
`;
